3836/3836.md

```diff
@@ -0,0 +1,666 @@
+```hs
+if full then error failureMsg
+```
+
+This error is triggered when trying to write to a full buffer. This behavior is
+intentional: the enclosing function makes its potentially partial semantics
+clear in its documentation, and declares itself "primarily useful for testing
+protocols." No changes are warranted.
+
+**Fix: none**
+#fixed
+
+
+## Tier 2
+
+Tier 2 files contain 11 uses of `error`.
+
+### 1: `cardano-ping/src/Cardano/Network/Ping.hs:183`
+```hs
+handshakeReqEnc [] _ = error "null version list"
+```
+
+The error is triggered when `handshakeReqEnc` is provided an empty list of
+`NodeVersion` - but its only caller, `handshakeReq`, also guards against
+providing an empty list. I've changed `handshakeReqEnc` to expect a `NonEmpty
+NodeVersion` to eliminate the error.
+
+**Fix: refactor**
+#fixed
+
+
+### 2: `network-mux/src/Network/Mux/DeltaQ/TraceStats.hs:37`
+```hs
+Nothing -> error "step: missing referenceTimePoint"
+```
+
+The error is triggered when the `referenceTimePoint` is `Nothing` - but this
+eventuality is handled in a prior function guard, so the error is never
+triggered. More careful pattern-matching eliminates the call to `error`.
+
+**Fix: refactor**
+#fixed
+
+
+### 3: `network-mux/src/Network/Mux/DeltaQ/TraceStats.hs:253`
+```hs
+= error "Infeasible sampleInterval"
+```
+
+This error is triggered when a developer tries to provide an unsuitable sample
+interval. The existing interval is suitable, so I assume this variable is
+somewhat volatile during some development process. The suitability of the
+provided interval can be determined statically, so I've lifted this computation
+to be compile-time via Template Haskell, so that the package fails to compile
+with an unsuitable interval.
+
+TH is polarizing, I realize, so I would be happy to discuss this in greater
+detail and revert it if there's no appetite for such a change.
+
+**Fix: refactor**
+#fixed
+
+#input-desired
+
+
+### 4: `network-mux/src/Network/Mux/Ingress.hs:185`
+```hs
+Nothing -> error ("setupDispatchTable: missing " ++ show miniProtocolNum)
+```
+
+The error occurs if, in the building of an array (`ptclArray`), an intermediate
+array (`pnumArray`) resolves an index (`miniProtocolNum`) to an "empty"
+(represented as `Nothing`) slot.
+
+`pnumArray` is constructed such that every `miniProtocolNum` in `ptcls` maps to
+a `Just`-constructed slot. `ptclArray` is constructed by using every
+`miniProtocolNum` in `ptcls` to index `pnumArray`, and will trigger this error
+on encountering `Nothing` as a result of that indexing. Therefore, by
+construction, the error will never trigger.
+
+Leveraging this principle, I've reworked the list comprehension to extract the
+`Just`-constructed value required using a view pattern. The semantics of this
+are now slightly different: if `pnumArray` fails to index, `ptclArray` will
+silently skip the offending element instead of loudly failing. I've included
+commentary to this effect.
```

Comment:
I'd prefer to avoid a view pattern, an explicit pattern match is more readable in my opinion.

Reply:
Reverted.